File: backend/tests/source_text.py
"""One helper for reading source as CODE, without its prose.

═══ WHY THIS FILE EXISTS ═══

EIGHT times now, a forbidden-pattern pin has failed on the comment
EXPLAINING the removal it was guarding. The shape is always the same:

    # This used to apply a generic $2.20 to any city at all.
    ...
    assert "2.20" not in src   # <- trips on the sentence above

The pin is right, the code is right, and the test fails because prose
that describes a defect necessarily quotes it. Every occurrence got its
own local strip, and by T-2 there were four of them in this directory
with slightly different behaviour — one stripped comments but not
docstrings, which is exactly how the sixth trip happened.

T-3 consolidated the Python side. The eighth trip (RED-H1.4, on a
comment recording an `OPENAI_AVAILABLE` removal) cost one import instead
of a debugging session, which is the argument for this file.

═══ WHAT CHANGED, AND WHY THE OLD VERSION WAS NOT ENOUGH ═══

The first implementation did:

    for node in ast.walk(tree):
        doc = ast.get_docstring(node)
        if doc:
            src = src.replace(doc, "")     # <- text replace

That is a substring replace over the WHOLE FILE. Two consequences:

  1. If a docstring's text also appears in code — a SQL fragment, an
     error message, a status literal — the replace deletes it THERE too,
     and a pin then passes because the thing it was guarding vanished
     from the text it searched. A pin that passes for the wrong reason is
     worse than one that fails.
  2. It removed the docstring's CONTENT but left its quotes, and took the
     newlines with it, so every line number below shifted and a failure
     message pointed at the wrong line.

It also only stripped comments on lines that START with `#`, so a
trailing `x = 1  # ...` survived and could trip a pin.

This version tokenizes. `tokenize` knows which `#` is a comment and which
is inside a string, and it reports exact positions, so comments and
docstrings are blanked IN PLACE. Line and column numbers survive.

Same lesson as everywhere else in this project, in its narrowest form:
match STATEMENTS, not strings. A tokenizer knows what a comment is; a
pattern only knows what one looks like.

═══ RANGE-BASED PINS ARE THE SAME FAMILY ═══

Owner-ruled after the eighth trip (PARTNER1). The rule reads as being
about forbidden STRINGS, and every earlier occurrence was one — a price,
a claim, a removed symbol name. It is not limited to strings.

PARTNER1's brand pass forbade emoji on a screen that had shipped them as
button labels. The pin was a CHARACTER RANGE:

    expect(RAW).not.toMatch(/[\\u{1F300}-\\u{1FAFF}\\u{2190}-\\u{27BF}]/u)

and it failed — on the ═══ rule characters in its own header comment,
because U+2500–U+257F (box drawing) sits inside U+2190–U+27BF. Same
shape as all eight: the prose describing the forbidden thing contains
the forbidden thing. Nothing about it being a range rather than a
literal changed that.

So the rule generalises: **any pin whose pattern could match decorative
or explanatory text runs against `code_only()` output, not raw source.**
Strings, regexes, character classes, Unicode ranges — the question is
only whether the pattern can hit a comment, and a wide range can hit far
more of one than a literal can.

Two corollaries worth stating, because the range case has them and the
string case does not:

  - Keep the range no wider than the thing you are forbidding. The emoji
    pin wanted emoji and dingbats; it asked for arrows, maths, box
    drawing and geometric shapes as well, and got what it asked for.
  - A pin that must read prose (checking a header cites a doctrine
    section, say) is a different question and should read RAW
    deliberately, saying so at the call site.

═══ WHAT IT DOES NOT DO ═══

It does not strip ordinary string literals. A pin searching for a value
that legitimately appears in a STRING (an error message, a SQL fragment)
is asking a different question and should say so at the call site.
"""
import io
import tokenize
from pathlib import Path
from typing import Union


def _blank_span(lines, start, end):
    """Replace the source between two (row, col) points with spaces,
    keeping newlines so every later position stays where it was."""
    srow, scol = start
    erow, ecol = end
    if srow == erow:
        line = lines[srow - 1]
        lines[srow - 1] = line[:scol] + " " * (ecol - scol) + line[ecol:]
        return
    first = lines[srow - 1]
    lines[srow - 1] = first[:scol] + " " * (len(first) - scol)
    for row in range(srow + 1, erow):
        lines[row - 1] = " " * len(lines[row - 1])
    last = lines[erow - 1]
    lines[erow - 1] = " " * ecol + last[ecol:]


def code_only(source: Union[str, Path]) -> str:
    """Return `source` with docstrings and `#` comments blanked out.

    Accepts a path or the source text itself. Positions are preserved:
    a comment becomes spaces, not nothing, so `line N` in a failure
    message still refers to line N of the real file.

    Falls back to the raw text if the source does not tokenize — a test
    helper must never turn a syntax error into a confusing assertion
    failure somewhere else.
    """
    src = Path(source).read_text(encoding="utf-8") if isinstance(source, Path) else source

    # keepends so column offsets line up with the original exactly.
    lines = src.splitlines(keepends=True)
    if not lines:
        return src

    try:
        tokens = list(tokenize.generate_tokens(io.StringIO(src).readline))
    except (tokenize.TokenError, IndentationError, SyntaxError):
        return src

    # A STRING token is a docstring when it is the first meaningful token
    # of a logical line — which covers module, class and function
    # docstrings AND the bare string-expression case, without ast's
    # node walk or its whole-file text replace.
    at_statement_start = True
    spans = []
    for tok in tokens:
        if tok.type == tokenize.COMMENT:
            spans.append((tok.start, tok.end))
            continue
        if tok.type in (tokenize.NL, tokenize.NEWLINE, tokenize.INDENT,
                        tokenize.DEDENT, tokenize.ENCODING):
            # ONLY the logical NEWLINE starts a new statement. NL is the
            # non-logical newline tokenize emits INSIDE brackets, and
            # treating it as a statement boundary was a real bug caught
            # by the existing suite: in
            #
            #     for stmt in [
            #         \"\"\"CREATE TABLE ...\"\"\",
            #
            # the SQL follows `[` + NL, so resetting here marked it a
            # docstring and blanked the entire schema out of database.py.
            # Ten pins failed at once — the over-stripping direction,
            # which is the one that makes pins pass for the wrong reason.
            if tok.type == tokenize.NEWLINE:
                at_statement_start = True
            continue
        if tok.type == tokenize.STRING and at_statement_start:
            spans.append((tok.start, tok.end))
        at_statement_start = False

    for start, end in reversed(spans):
        _blank_span(lines, start, end)

    return "".join(lines)


def read_code(*segments: str) -> str:
    """Read a file relative to `backend/` and return it as code only."""
    return code_only(Path(__file__).resolve().parent.parent.joinpath(*segments))

File: frontend/src/test-support/sourceText.ts
/**
 * codeOnly() — read source as CODE, without its prose. The TS half.
 *
 * ═══ WHY THIS EXISTS ═══
 *
 * A forbidden-pattern pin necessarily quotes what it forbids, and the
 * comment explaining a removal necessarily quotes the thing removed. So
 * every such pin has to read code with comments stripped, and this repo
 * grew SIXTEEN inline strippers across fourteen test files to do it.
 *
 * They did not agree, and two of the three variants were wrong:
 *
 *   /\/\/[^\n]*​/g          (6 files)
 *       Strips `//` ANYWHERE — including inside string literals. It turns
 *
 *           const url = "https://api.openai.com/v1/chat";
 *       into
 *           const url = "https:
 *
 *       It truncates the string, leaves an unterminated quote, and
 *       deletes the rest of the line. Several files using it test proxy
 *       and docs code whose whole content is URLs, so those pins have
 *       been asserting against mangled text.
 *
 *   /^\s*\/\/.*$​/gm        (6 files)
 *       `\s` matches newlines, so a blank line before a comment is eaten
 *       WITH the comment and every line number below shifts. Same bug
 *       found independently in the banned-claims checker (RED-H1.1).
 *
 *   /^[^\S\n]*\/\/.*$​/gm   (1 file)
 *       Correct, and the reason the other two are now gone.
 *
 * ═══ WHY A SCANNER AND NOT A BETTER REGEX ═══
 *
 * "Is this `//` a comment or part of a string?" is not a question a
 * regular expression can answer — it requires knowing what came before.
 * That is the same lesson this project keeps relearning under a
 * different name: match STATEMENTS, not strings. A tokenizer knows the
 * difference between a comment and text that looks like one; a pattern
 * only knows the spelling.
 *
 * ═══ WHAT IT DOES NOT DO ═══
 *
 * It does not strip string literals — a pin looking for a value that
 * legitimately lives in a string is asking a different question and
 * should say so at the call site.
 *
 * It does not fully parse regex literals. `/abc/.test(s)` is handled by
 * the division-vs-regex heuristic below; a regex literal containing an
 * unbalanced quote could still confuse it. That is stated rather than
 * hidden: if it ever matters, the fix is a real parser, not a patch.
 *
 * Line and column positions are PRESERVED — comments become spaces,
 * newlines stay newlines — so a failing assertion's line numbers still
 * point at the right place in the original file.
 */
import * as fs from 'fs';

/** Characters that can precede `/` when it starts a regex literal. */
const REGEX_ALLOWED_BEFORE = new Set([
  '(', ',', '=', ':', '[', '!', '&', '|', '?', '{', '}', ';', '\n', '+', '-', '*', '%', '<', '>', '~', '^',
]);

function lastSignificant(out: string[]): string {
  for (let i = out.length - 1; i >= 0; i--) {
    const c = out[i];
    if (c !== ' ' && c !== '\t' && c !== '\r') return c;
  }
  return '\n';
}

export function codeOnly(source: string): string {
  const out: string[] = [];
  let i = 0;
  const n = source.length;

  // Blank a span while keeping newlines, so positions never move.
  const blank = (from: number, to: number) => {
    for (let k = from; k < to; k++) out.push(source[k] === '\n' ? '\n' : ' ');
  };

  while (i < n) {
    const c = source[i];
    const next = source[i + 1];

    // Line comment
    if (c === '/' && next === '/') {
      let j = i;
      while (j < n && source[j] !== '\n') j++;
      blank(i, j);
      i = j;
      continue;
    }

    // Block comment (this also covers JSX `{/* ... */}`)
    if (c === '/' && next === '*') {
      let j = i + 2;
      while (j < n && !(source[j] === '*' && source[j + 1] === '/')) j++;
      j = Math.min(j + 2, n);
      blank(i, j);
      i = j;
      continue;
    }

    // String literal — copied verbatim, so `https://` survives intact.
    if (c === '"' || c === "'") {
      out.push(c);
      i++;
      while (i < n) {
        if (source[i] === '\\') { out.push(source[i], source[i + 1] ?? ''); i += 2; continue; }
        out.push(source[i]);
        if (source[i] === c) { i++; break; }
        if (source[i] === '\n') { i++; break; } // unterminated — don't run away
        i++;
      }
      continue;
    }

    // Template literal, including nested ${ ... } which may itself hold
    // strings and comments.
    if (c === '`') {
      out.push(c);
      i++;
      let depth = 0;
      while (i < n) {
        if (source[i] === '\\') { out.push(source[i], source[i + 1] ?? ''); i += 2; continue; }
        if (source[i] === '$' && source[i + 1] === '{') { depth++; out.push('$', '{'); i += 2; continue; }
        if (source[i] === '}' && depth > 0) { depth--; out.push('}'); i++; continue; }
        if (source[i] === '`' && depth === 0) { out.push('`'); i++; break; }
        out.push(source[i]);
        i++;
      }
      continue;
    }

    // Regex literal, distinguished from division by what precedes it.
    if (c === '/' && REGEX_ALLOWED_BEFORE.has(lastSignificant(out))) {
      let j = i + 1;
      let inClass = false;
      let closed = false;
      while (j < n) {
        if (source[j] === '\\') { j += 2; continue; }
        if (source[j] === '[') inClass = true;
        else if (source[j] === ']') inClass = false;
        else if (source[j] === '/' && !inClass) { closed = true; j++; break; }
        else if (source[j] === '\n') break;
        j++;
      }
      if (closed) {
        for (let k = i; k < j; k++) out.push(source[k]);
        i = j;
        continue;
      }
    }

    out.push(c);
    i++;
  }

  return out.join('');
}

/** Read a file under `frontend/src` and return it as code only. */
export function readCode(...segments: string[]): string {
  const path = require('path');
  return codeOnly(fs.readFileSync(path.join(__dirname, '..', '..', ...segments), 'utf8'));
}

/** Read a file relative to the repository root and return it as code only. */
export function readRepoCode(...segments: string[]): string {
  const path = require('path');
  return codeOnly(
    fs.readFileSync(path.join(__dirname, '..', '..', '..', '..', ...segments), 'utf8')
  );
}
